File: Algorithms/SearchingAlgorithms/HillClimbing/java/HillClimbing.java
import java.util.Random;

public class HillClimbing {
    private static final Random random = new Random();

    public static void main(String[] args) {
        double[] result = hillClimbing(1000, 0.1);
        System.out.printf("Best x: %.2f, Best value: %.2f%n", result[0], result[1]);
    }

    public static double[] hillClimbing(int maxIter, double stepSize) {
        // Start at a random x
        double currentX = random.nextDouble() * 20 - 10; // Range: [-10, 10]
        double currentVal = -Math.pow(currentX, 2) + 4 * currentX;

        for (int i = 0; i < maxIter; i++) {
            // Generate a neighboring solution
            double neighborX = currentX + (random.nextDouble() * 2 - 1) * stepSize;
            double neighborVal = -Math.pow(neighborX, 2) + 4 * neighborX;

            // Move to the neighbor if it's better
            if (neighborVal > currentVal) {
                currentX = neighborX;
                currentVal = neighborVal;
            }
        }

        return new double[]{currentX, currentVal};
    }
}

File: Algorithms/SearchingAlgorithms/HillClimbing/py/Hillclimbing.py
import random

def hill_climbing(max_iter=1000, step_size=0.1):
    # Start at a random x
    current_x = random.uniform(-10, 10)
    current_val = -current_x**2 + 4*current_x
    
    for _ in range(max_iter):
        # Generate a neighboring solution
        neighbor_x = current_x + random.uniform(-step_size, step_size)
        neighbor_val = -neighbor_x**2 + 4*neighbor_x
        
        # Move to the neighbor if it's better
        if neighbor_val > current_val:
            current_x, current_val = neighbor_x, neighbor_val
    
    return current_x, current_val

# Run Hill Climbing
best_x, best_val = hill_climbing()
print(f"Best x: {best_x:.2f}, Best value: {best_val:.2f}")